### ISSUE TITLE
Roslyn Code Actions violate language server protocol by making text changes server => client, corrupting semantic tokens

### ISSUE THREAD

**Issue by gundermanc** [dotnet/roslyn#48991]
**Version Used**: 
Microsoft Visual Studio Enterprise 2019 main
Version 16.9.0 Preview 2.0 [30627.110.main]

**Steps to Reproduce**:

1. Open a C# document
2. Wait for semantic colorization to appear
3. Invoke a LightBulb action that causes a code change in the current document.

**Expected Behavior**:
Change is applied on the client without disrupting colorization.

**Actual Behavior**:
Change is applied on the server and remoted to the client via coauthoring, enabling semantic tokens requests to be queued against some client version _n_, but actually run against _n+1_, where 1 is a version created on the server and dispatched to the client.

This leads to a corruption of the visible colors in the document (shifts all colorized spans a few characters left or right) or triggers a refresh of the entire document's colorization due to out of bounds colorized spans (very expensive).

**Comment by CyrusNajmabadi** [comment]
tagging @davidwengier @dibarbet @allisonchou 

> Change is applied on the server 

I think this is a requirement for us.  Because only we know the set of all changes to make to a solution.  The client won't know all this, and we have no way to convey to them all the changes to make.

**Comment by gundermanc** [comment]
Tagging @olegtk as this is closely related to refactors we are looking at now in LSP.

**Comment by dibarbet** [comment]
@CyrusNajmabadi @gundermanc 
It seems like it would be possible, at least in particular instances to apply the text changes only to open documents via workspace edits and then create commands to do things like adding new files / adding project references (in effect, split the code action into workspace edit vs command by what the client has open).
It is allowed to return workspace edits and commands per the LSP spec.  The workspace edits will be applied first on the client, then the client calls the server to execute the command.

Perhaps extending this further, maybe we should _only_ send back workspace edits for files the client has open, the server can apply everything else.  This avoids the client needing to download and open 100 files if for example you rename an identifier.  And per the LSP spec, the client only owns open files, so we should be able to do whatever we want to closed files.

**Comment by gundermanc** [comment]
@dibarbet I like the idea of coming up with a standard, well-designed solution for this problem. The one you suggested seems reasonable to me, if we can guarantee synchronization with the change distributed between client and server like that.

Another alternative is that we could attempt to make the LSP client on the server a bit smarter: make the LSP client have special understanding of the concept of server-side modifications, gating any _textDocument/didChange_ notifications from devenv server on receiving an ACK by all clients.

One problem with the current text synchronization model is that it is operating more or less independently of LSP. Since each LSP request is versioned implicitly by text order, this means that there are some latent race conditions in features like colorization and completion that have had to be fixed on a case by case basis.

This doc explains the problem and some potential redesigns for LSP messages and text synchronization: https://microsoft.sharepoint.com/:w:/t/MS-LSP2/EdeqP_IoJvlJgnALJeCsh-sBeXZfpaFAqiYgCR3e_RL4jg?e=LoHkfV

**Comment by CyrusNajmabadi** [comment]
> It seems like it would be possible, at least in particular instances to apply the text changes only to open documents via workspace edits and then create commands to do things like adding new files / adding project references (in effect, split the code action into workspace edit vs command by what the client has open).

This might be possible in constrained cases, but not in the general cast.  For example, in Roslyn we allow for arbitrary workspace edits with codeactions.  And there is no way to introspect an arbitrary action to know what it will do until it actually goes and does it.  We also have a contract that says any of our code-actoins (or third part actions) can change their impl in this fashion and it will not break scenarios.

This LSP limitation breaks that.  IMO, if LSP supports the concept of code-action-commands, then that has to be understood and baked into the protocol for other delta-oriented APIs.  There's simply no way to isolate the host from arbitrary doc edits being made, so any sort of sync'ing protocol needs to account for those changes happening IMO.  

In other words.  We could attempt o fix this for a narrow set of cases.  But the overall issue would remain for other code actions, and for other host initiated edits.  

Given this, i think we should roll back our semantic-tokens work until we can come up with a design where it works for all the scenarios where we want to use LSP today and in the future.

**Comment by CyrusNajmabadi** [comment]
> https://microsoft.sharepoint.com/:w:/t/MS-LSP2/EdeqP_IoJvlJgnALJeCsh-sBeXZfpaFAqiYgCR3e_RL4jg?e=LoHkfV

This link is not available.  Can you host this on something public?  Or through devdiv.visualstudio.com?

**Comment by davidwengier** [comment]
@gundermanc In theory when we start responding to `didChange` this problem should be solved, as the semantic classification request will be operating on the right document state. If you want to validate that you can install a build with that enabled in Roslyn here: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4185062&view=ms.vss-build-web.run-extensions-tab

**Comment by gundermanc** [comment]
Thanks, I'll give that a try. I also noticed that the particular example I gave in the repro here is using a DocumentEdit sent to the client, so this particular case is probably actually an LSP client bug.

There might still be other bugs in Roslyn though if there are any cases where the change flows server => client through coauthoring, so it's probably worth thinking through this scenario regardless.

**Comment by davidwengier** [comment]
> I also noticed that the particular example I gave in the repro here is using a DocumentEdit sent to the client,

That's good to know. There are definitely code actions that don't send workspace edits, so we know this problem exists in Roslyn, but we're not _entirely_ bad :)

**Comment by gundermanc** [comment]
FYI: in response to our conversation last Friday, I have filed a proposed user story to harden the LSP client's portion of these scenarios: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1243527.

The intent is that if possible, we support server-side edits and have a 'gating' mechanism where we don't actually inform the LSP client about the edit until it has been acknowledged by the client.

I don't have any more details yet, this will probably require some prototyping.

**Comment by dibarbet** [comment]
Closing to be tracked by https://github.com/dotnet/roslyn/issues/48991

**Comment by allisonchou** [comment]
@dibarbet did you mean to link a different issue?

**Comment by dibarbet** [comment]
yes, updated.  thanks for the catch!
